### PR TITLE
fix: slow sync stuck resolving 1-1s WPB-6861

### PR DESCRIPTION
**Pull request by johnxnguyen**
----
#### PR Submission Checklist for internal contributors

- The **PR Title**
  - [x] conforms to the style of semantic commits messages¹ supported in Wire's Github Workflow²
  - [x] contains a reference JIRA issue number like `SQPIT-764`
  - [x] answers the question: _If merged, this PR will: ..._ ³

- The **PR Description**
  - [x] is free of optional paragraphs and you have filled the relevant parts to the best of your ability
----

# What's new in this PR?

### Issues

When logging in, the slow sync gets stuck resolving 1-1 conversations indefinitely. Restarting the app doesn't help.

### Causes

If the user logging in has many 1-1s, then there are many 1-1s to resolve. Resolving a conversation may result in several network requests, for instance to fetch the mls one conversation. We were attempting to resolve each 1-1 in parallel within a task group, which resulted in a flood of network requests to the same endpoint (one for each conversation) in a very short amount of time. Quickly the BE returned back error responses with code 420 indicating that we hit rate limits. This code is not handled by the action handlers and the phase to resolve 1-1s failed but did not (for some reason) end the slow sync.

### Solutions

~The simplest solution is to resolve the 1-1s serially. This avoids the rate limits, but is fairly slow.~

When the action handler receives a rate limit error, push the action back in the list of pending actions. This ensures that the next time the operation loop runs again (after a backoff, handled elsewhere) it can try to perform the action again.

Another issue was found where the transport session tries to notify that new requests are available after backing off but the operation loop never starts again. This was because it was firing a "ZMReqestAvailableNotification" notification but the operation loop listens for "RequestAvailableNotification". I got rid of the former so there's only one notification.

#### How to Test

1. Log in with an MLS capable client with an account that has many contacts (either a large team or many accepted connection requests.
2. Assert that the slow sync finishes and the 1-1s are resolved.

----
#### PR Post Submission Checklist for internal contributors (Optional)

 - [ ] Wire's Github Workflow has automatically linked the PR to a JIRA issue
----
#### PR Post Merge Checklist for internal contributors

 - [ ] If any soft of configuration variable was introduced by this PR, it has been added to the relevant documents and the CI jobs have been updated.
----
##### References
1. https://sparkbox.com/foundry/semantic_commit_messages
1. https://github.com/wireapp/.github#usage
1. E.g. `feat(conversation-list): Sort conversations by most emojis in the title #SQPIT-764`.

### CODE REVIEW

**Review by typfel — APPROVED** [review]

File: wire-ios-request-strategy/Sources/Entity Actions/ActionHandler.swift [hunk 1]
```diff
@@ -58,7 +58,13 @@ class ActionHandler<T: EntityAction>: NSObject, EntityActionHandler, ZMRequestGe
         let request = self.request(for: action, apiVersion: apiVersion)
 
         request?.add(ZMCompletionHandler(on: context, block: { [weak self] (response) in
-            self?.handleResponse(response, action: action)
+            if response.httpStatus.isOne(of: TooManyRequestsStatusCode, EnhanceYourCalmStatusCode) {
+                // We're being rate limited, put the action back so we can try again
+                // next time the operation loop polls.
```
Comment (netbe) [inline wire-ios-request-strategy/Sources/Entity Actions/ActionHandler.swift:63]: 🤔 pulls or polls but again I am not native speaker
```suggestion
                // next time the operation loop pulls.
```

**Review by netbe — APPROVED** [review]
LGTM;) just a small comment

**Review by typfel — APPROVED** [review]
Nice! Not that much changed!

File: wire-ios-request-strategy/Sources/Entity Actions/ActionHandler.swift (hunk 1 above)
Comment (typfel) [reply wire-ios-request-strategy/Sources/Entity Actions/ActionHandler.swift:63]: I think polling is the correct term

**Comment by typfel** [comment]
@johnxnguyen does it take a long to time to recover after you've been rate limited?

**Comment by johnxnguyen** [comment]
> @johnxnguyen does it take a long to time to recover after you've been rate limited?

@typfel yes about 1min 20sec for an account with 94 one on one conversations, and we get limited about 5 times.

File: wire-ios-request-strategy/Sources/Entity Actions/ActionHandler.swift (hunk 1 above)
Comment (johnxnguyen) [reply wire-ios-request-strategy/Sources/Entity Actions/ActionHandler.swift:63]: Polls in the correct term

**Pull request merged**